File: java/src/kanzi/test/TestBWT.java
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import java.util.Random;
import kanzi.IndexedByteArray;
import kanzi.transform.BWT;


public class TestBWT
{
    public static void main(String[] args)
    {
        System.out.println("TestBWT");
        testCorrectness();
        testSpeed();
    }
    
    
    public static void testCorrectness()
    {
        System.out.println("\nCorrectness test");

        // Test behavior
        for (int ii=1; ii<=20; ii++)
        {
            System.out.println("\nTest "+ii);
            int start = 0;
            int size = 0;
            byte[] buf1;
            Random rnd = new Random();

            if (ii == 1)
            {
               size = 0;
               buf1 = "mississippi".getBytes();
            }
            else if (ii == 2)
            {
               buf1 = "3.14159265358979323846264338327950288419716939937510".getBytes();
            }
            else
            {
               size = 128;
               buf1 = new byte[size];

               for (int i=0; i<buf1.length; i++)
               {
                   buf1[i] = (byte) (65 + rnd.nextInt(4*ii));
               }

               buf1[buf1.length-1] = (byte) 0;
            }

            byte[] buf2 = new byte[buf1.length];
            byte[] buf3 = new byte[buf1.length];
            IndexedByteArray iba1 = new IndexedByteArray(buf1, 0);
            IndexedByteArray iba2 = new IndexedByteArray(buf2, 0);
            IndexedByteArray iba3 = new IndexedByteArray(buf3, 0);
            BWT bwt = new BWT(size);
            String str1 = new String(buf1, start, buf1.length-start);
            System.out.println("Input:   "+str1);
            iba1.index = start;
            iba2.index = 0;
            bwt.forward(iba1, iba2);
            int primaryIndex = bwt.getPrimaryIndex();
            String str2 = new String(buf2);
            System.out.print("Encoded: "+str2);
            System.out.println("  (Primary index="+primaryIndex+")");
            bwt.setPrimaryIndex(primaryIndex);
            iba2.index = 0;
            iba3.index = start;
            bwt.inverse(iba2, iba3);
            String str3 = new String(buf3, start, buf3.length-start);
            System.out.println("Output:  "+str3);

            if (str1.equals(str3) == true)
               System.out.println("Identical");
            else
            {
               System.out.println("Different");
               System.exit(1);
            }
        }
    }
    
    
    public static void testSpeed()
    {
         System.out.println("\nSpeed test");
         int iter = 2000;
         int size = 256*1024;
         byte[] buf1 = new byte[size];
         byte[] buf2 = new byte[size];
         byte[] buf3 = new byte[size];
         IndexedByteArray iba1 = new IndexedByteArray(buf1, 0);
         IndexedByteArray iba2 = new IndexedByteArray(buf2, 0);
         IndexedByteArray iba3 = new IndexedByteArray(buf3, 0);
         System.out.println("Iterations: "+iter);
         System.out.println("Transform size: "+size);

         for (int jj = 0; jj < 3; jj++)
         {
             long delta1 = 0;
             long delta2 = 0;
             BWT bwt = new BWT(size);
             java.util.Random random = new java.util.Random();
             long before, after;

             for (int ii = 0; ii < iter; ii++)
             {
                 for (int i = 0; i < size; i++)
                     buf1[i] = (byte) (random.nextInt(255) + 1);

                 buf1[size-1] = 0;
                 before = System.nanoTime();
                 iba1.index = 0;
                 iba2.index = 0;
                 bwt.forward(iba1, iba2);
                 after = System.nanoTime();
                 delta1 += (after - before);
                 before = System.nanoTime();
                 iba2.index = 0;
                 iba3.index = 0;
                 bwt.inverse(iba2, iba3);
                 after = System.nanoTime();
                 delta2 += (after - before);
             
               int idx = -1;

               // Sanity check
               for (int i=0; i<size; i++)
               {
                  if (buf1[i] != buf3[i])
                  {
                     idx = i;
                     break;
                  }
               }

               if (idx >= 0)
                  System.out.println("Failure at index "+idx+" ("+buf1[idx]+"<->"+buf3[idx]+")");             
             }

            final long prod = (long) iter * (long) size;
            System.out.println("Forward transform [ms] : " + delta1 / 1000000);
            System.out.println("Throughput [KB/s]      : " + prod * 1000000L / delta1 * 1000L / 1024);
            System.out.println("Inverse transform [ms] : " + delta2 / 1000000);
            System.out.println("Throughput [KB/s]      : " + prod * 1000000L / delta2 * 1000L / 1024);
         }
    }
}

File: go/src/kanzi/test/TestBWT.go
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"kanzi"
	"kanzi/transform"
	"math/rand"
	"os"
	"time"
)

func main() {
	var input = flag.String("input", "", "input string")

	// Parse
	flag.Parse()

    if len(*input) > 0 {
    	buf1 := []byte(*input)
    	size := uint(len(buf1))
    	buf2 := make([]byte, size)
    	bwt, _ := transform.NewBWT(size)
    	bwt.Forward(buf1, buf2)
		fmt.Printf("BWT:  %s (%v)\n", buf2, bwt.PrimaryIndex())
    	bwts, _ := transform.NewBWTS(size)
    	bwts.Forward(buf1, buf2)
		fmt.Printf("BWTS: %s\n", buf2)
		os.Exit(0)
    }
        
	fmt.Printf("TestBWT and TestBWTS")
	TestCorrectness(true)
	TestCorrectness(false)
	TestSpeed(true)
	TestSpeed(false)
}

func TestCorrectness(isBWT bool) {
	if isBWT {
		fmt.Printf("\n\nBWT Correctness test")
	} else {
		fmt.Printf("\n\nBWTS Correctness test")
	}

	// Test behavior
	for ii := 1; ii <= 20; ii++ {
		fmt.Printf("\nTest %v\n", ii)
		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))

		size := uint(0)
		var buf1 []byte
		var buf2 []byte
		var buf3 []byte

		if ii == 1 {
			size = 0
			buf1 = []byte("mississippi")
		} else if ii == 2 {
			buf1 = []byte("3.14159265358979323846264338327950288419716939937510")
		} else {
			size = 128
			buf1 = make([]byte, size)

			for i := 0; i < len(buf1); i++ {
				buf1[i] = byte(65 + rnd.Intn(4*ii))
			}
		}

		buf2 = make([]byte, len(buf1))
		buf3 = make([]byte, len(buf1))
		var bwt kanzi.ByteTransform

		if isBWT {
			bwt, _ = transform.NewBWT(size)
		} else {
			bwt, _ = transform.NewBWTS(size)
		}

		str1 := string(buf1)
		fmt.Printf("Input:   %s\n", str1)
		_, _, err1 := bwt.Forward(buf1, buf2)

		if err1 != nil {
			fmt.Printf("Error: %v\n", err1)
			os.Exit(1)
		}

		str2 := string(buf2)
		fmt.Printf("Encoded: %s", str2)

		if isBWT {
			primaryIndex := bwt.(*transform.BWT).PrimaryIndex()
			fmt.Printf("  (Primary index=%v)\n", primaryIndex)
			bwt.(*transform.BWT).SetPrimaryIndex(primaryIndex)
		} else {
			println()
		}
		
		_, _, err2 := bwt.Inverse(buf2, buf3)

		if err2 != nil {
			fmt.Printf("Error: %v\n", err2)
			os.Exit(1)
		}

		str3 := string(buf3)
		fmt.Printf("Output:  %s\n", str3)

		if str1 == str3 {
			fmt.Printf("Identical\n")
		} else {
			fmt.Printf("Different\n")
			os.Exit(1)
		}
	}
}

func TestSpeed(isBWT bool) {
	if isBWT {
		fmt.Printf("\n\nBWT Speed test")
	} else {
		fmt.Printf("\n\nBWTS Speed test")
	}

	iter := 2000
	size := 256 * 1024
	buf1 := make([]byte, size)
	buf2 := make([]byte, size)
	buf3 := make([]byte, size)
	fmt.Printf("\nIterations: %v", iter)
	fmt.Printf("\nTransform size: %v\n", size)

	for jj := 0; jj < 3; jj++ {
		delta1 := int64(0)
		delta2 := int64(0)
		var bwt kanzi.ByteTransform

		if isBWT {
			bwt, _ = transform.NewBWT(0)
		} else {
			bwt, _ = transform.NewBWTS(0)
		}

		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))

		for i := 0; i < iter; i++ {
			for i := range buf1 {
				buf1[i] = byte(rnd.Intn(255) + 1)
			}

			before := time.Now()
			bwt.Forward(buf1, buf2)
			after := time.Now()
			delta1 += after.Sub(before).Nanoseconds()
			before = time.Now()
			bwt.Inverse(buf2, buf3)
			after = time.Now()
			delta2 += after.Sub(before).Nanoseconds()

			// Sanity check
			for i := range buf1 {
				if buf1[i] != buf3[i] {
					fmt.Printf("Error at index %v: %v<->%v\n", i, buf1[i], buf3[i])
					os.Exit(1)
				}
			}
		}

		println()
		prod := int64(iter) * int64(size)
		fmt.Printf("Forward transform [ms] : %v\n", delta1/1000000)
		fmt.Printf("Throughput [KB/s]      : %d\n", prod*1000000/delta1*1000/1024)
		fmt.Printf("Inverse transform [ms] : %v\n", delta2/1000000)
		fmt.Printf("Throughput [KB/s]      : %d\n", prod*1000000/delta2*1000/1024)
		println()
	}
}
